### ISSUE TITLE
Closed Caption support?

### ISSUE THREAD

**Issue by localpcguy** [brianegan/chewie#77]
Does Chewie (or video_player) support Closed Captions? 

It is a requirement for us (and generally something that a video plugin should probably support for good accessibility). 

I wasn't sure if I should start here or if that would be a feature of video_player itself and I should be asking over on that repo.

**Comment by snaeji** [comment]
I was looking at the video plugin and from my understanding the video plugin doesn't currently support closed captions.

**Comment by localpcguy** [comment]
I've added an issue for the `video_player` plugin, if anyone wants to add comments there:
https://github.com/flutter/flutter/issues/25388

From a UI perspective in Chewie, it could potentially need a CC button and control menu added to the controls bar.

**Comment by localpcguy** [comment]
I agree the UI itself (i.e. the CC button) would be good to implement in Chewie, but the basic functionality needs to be in the VideoPlayer plugin. Otherwise people who don't implement Chewie (i.e. those who need a custom UI) would have to implement CC themselves and it's not likely to happen.

**Comment by cbenhagen** [comment]
What would be that basic functionality? Would that include rendering the text on top of the Video? Where do we draw the line?

What about a completely separate CC plugin which can then be used by Chewie or other Players? It would not care about drawing, but just read the subtitles files and provide a `Listenable` with the text to be rendered? A `CcController` which you then could pass to Chewie?

Adding this external to the core plugins would also allow us to move much faster with this while reducing the workload of the core devs.

Customising the UI of Chewie already is and will be much easier in the near future.

**Comment by localpcguy** [comment]
I think it is really important that VideoPlayer at a minimum reads the subtitle files and provides a method to subscribe to the text to be rendered. I do not think that can be a separate plugin, particularly since it is really mostly exposing functionality in the underlying Android/iOS functionality. 

I'm not necessarily opposed to seeing the functionality in an external addon to get the functionality available if that's faster, but I firmly believe that functionality mentioned above should be part of the core plugin as soon as possible and then that plugin would be able to be deprecated. 

It should not specifically be tied to Chewie either.

**Comment by rakyat-game** [comment]
Please first update core plugins video_player first, look at
[This](https://github.com/flutter/flutter/issues/25388)

Try to update chewie for displaying purpose only one file:
- lib/src/material_controls.dart

How to use:
Before: `VideoPlayerController.network(videoUrl)`
Updated: `VideoPlayerController.network(videoUrl, subtitleUrl)`

File Attachment
[material_controls.zip](https://github.com/brianegan/chewie/files/3300495/material_controls.zip)

**Comment by snaeji** [comment]
I was able to build on @rakyat-game code and read VTT files and default to subtitles in the HLS stream.

The solution is rather dirty, language default is hardcoded and defaults to Icelandic subtitles if they are included in the HLS stream.
But this will maybe help somebody if he wants to use it for reference etc.

https://github.com/snaeji/video_player-0.10.1-3
https://github.com/snaeji/chewie-0.9.7

**Comment by cbenhagen** [comment]
@snaeji can you open a PR or review the one @kirill09 made? https://github.com/brianegan/chewie/pull/188

**Comment by snaeji** [comment]
Hey @cbenhagen !

I went over #188 and it looks pretty solid.

My implementation depends on changes to the video_player plugin. Exoplayer takes care of reading external or embedded subtitles and passes a list of available subtitles and what subtitles to display back to Chewie. It works on Android but i haven't looked at the IOS side and would definitely need some help with that or someone to take the lead.

@kirill09 #188 takes in a subtitle file as a string and displays the subtitle over the video. Not sure which format but probably srt. Implementation would not work with embedded subs.

So it really depends upon what direction we want to take. What is your take on this @cbenhagen ?

**Comment by kirill09** [comment]
hi @snaeji yes I use STR format. I forgot to make an example. 
I don't think the solution should depend on the platform.  video_player gives you the ability to show video of everything else better to write on dart

**Comment by localpcguy** [comment]
We are getting closer to the point were we are going to need this functionality, and we definitely want it to utilize embedded subtitles. We have experience implementing this in both iOS and Android natively in the past, so if we get to the point where we need this functionality and the Video_Player plugin isn't updated yet, we'll be submitting PRs to them to add in the functionality. 

I like the idea that for Chewie, the solution presented accepts the subs from Video_Player, regardless of whether they are embedded or standalone originally. That seems to be the most flexible and keeps Chewie as a UI library as opposed to needing to fetch/load subs itself.

**Comment by brianegan** [comment]
@localpcguy Completely agree with that strategy -- Chewie really is meant to just be the UI portion and the core logic would be best implemented in video player and the corresponding UI portion could be handled by Chewie.

Not sure if you're interested, but frankly I only use this project myself on the side and not in a professional project, and I'd love to get more maintainers on this project that are using it for real-life business purposes. Would you be interested in being added as a maintainer to this project so you can merge PRs and ship releases?

**Comment by snaeji** [comment]
I’ll be back from vacation in a week and then i’ll refine the code and submit a pr to chewie and the video_player plugin.

**Comment by Joran-Dob** [comment]
I now created my own package for this implementation in the dart, it wraps a video player and accepts a video player in its controller to now the current position. 

You can find it [here](https://pub.dev/packages/subtitle_wrapper_package), hope it helps

**Comment by tuanbs** [comment]
Based on [this example](https://github.com/flutter/plugins/blob/9859a383c7d15966eb9165f2616afeb4ee53129a/packages/video_player/video_player/example/lib/main.dart#L208), I am able to show subtitle using `Chewie` and the latest version of `Video Player` (0.10.7) which supports closed caption (`srt` format). But it doesn't show subtitle in `fullscreen` mode.
```dart
return AspectRatio(
  aspectRatio: _videoPlayerController.value.aspectRatio,
  child: Stack(
    alignment: Alignment.bottomCenter,
    children: <Widget>[
      Chewie(controller: _chewieController), // Use the VideoPlayer widget to display the video.
      ClosedCaption(
        text: _videoPlayerController.value.caption.text,
      ),
      // _PlayPauseOverlay(controller: _videoPlayerController),
      VideoProgressIndicator(_videoPlayerController, allowScrubbing: true),
    ],
  ),
);
```
Some other issues I found are: 
- The subtitle text is not aligned center.
- The subtitle's text size is not responsive. i.e it has a fixed size regardless of the width of video in portrait or landscape mode.

Any ideas or workaround is much appreciated. Thanks.

**Comment by mploigt** [comment]
This would be exactly what I need, but woking subtitles on full screen is essential.

**Comment by JohnGalt1717** [comment]
+1 with WebVTT support since that's what most browsers actually support natively.

**Comment by jvinai** [comment]
Any update on how to change subtitles on the fly?
I have a selector that's allowing the user to select languages for the subtitles but using `setSubtitle` doesnt seem to do the job...

**Comment by kvenn** [comment]
I've switched to https://github.com/itsSagarBro/modern_player which automatically supports subtitles and `mkv` file format.

It's just missing full screen support, which should be coming soon.
